container/lru/set_test.go

```diff
@@ -0,0 +1,671 @@
+	}, {
+		name:     "hit ratio 0% single item, expiration disabled",
+		numItems: 10,
+		ttl:      0,
+		itemIdxs: []uint64{10},
+		expected: []float64{0},
+	}, {
+		name:       "hit ratio 10% descending, expiration disabled, 3 would be",
+		numItems:   10,
+		numExpired: 3,
+		ttl:        0,
+		itemIdxs:   []uint64{0, 10, 11, 12, 13, 14, 15, 16, 17, 18},
+		expected:   []float64{100, 50, 33.33, 25, 20, 16.67, 14.29, 12.5, 11.11, 10},
+	}, {
+		name:       "hit ratio 50% alternating due to expired items",
+		numItems:   10,
+		numExpired: 5,
+		ttl:        expireScanInterval / 2,
+		itemIdxs:   []uint64{5, 0, 6, 1, 7, 2, 8, 3, 9, 4},
+		expected:   []float64{100, 50, 66.67, 50, 60, 50, 57.14, 50, 55.56, 50},
+	}, {
+		name:     "hit ratio 50% alternating, expiration disabled",
+		numItems: 10,
+		ttl:      0,
+		itemIdxs: []uint64{0, 10, 1, 11, 2, 12, 3, 13, 4, 14},
+		expected: []float64{100, 50, 66.67, 50, 60, 50, 57.14, 50, 55.56, 50},
+	}, {
+		name:       "hit ratio would be 100%, but 1 item expired",
+		ttl:        expireScanInterval / 2,
+		numItems:   5,
+		numExpired: 1,
+		itemIdxs:   []uint64{0, 1, 2, 3, 4},
+		expected:   []float64{0, 50, 66.67, 75, 80},
+	}, {
+		name:       "hit ratio 100%, expiration disabled, 5 would be expired",
+		ttl:        0,
+		numItems:   10,
+		numExpired: 5,
+		itemIdxs:   []uint64{0, 1, 2, 3, 4, 5, 6, 7, 8, 9},
+		expected:   []float64{100, 100, 100, 100, 100, 100, 100, 100, 100, 100},
+	}}
+
+	for _, test := range tests {
+		if len(test.itemIdxs) != len(test.expected) {
+			t.Fatalf("%q: test data item idx len mismatch: %d != %d", test.name,
+				len(test.itemIdxs), len(test.expected))
+		}
+		if test.numExpired > test.numItems {
+			t.Fatalf("%q: test data num expired exceeds num items: %d > %d",
+				test.name, test.numExpired, test.numItems)
+		}
+
+		// Create a new set with a TTL specified by the test that is big enough
+		// to hold all test values.  Also override the now function so the
+		// current time can be manipulated for forcing expirations.
+		mockNow = baseNow
+		set := NewSetWithDefaultTTL[uint64](totalItems+1, test.ttl)
+		set.m.nowFn = mockNowFn
+		set.m.nextExpireScan = baseNow.Add(expireScanInterval)
+		tc := &testContextSet{t: t, name: test.name, s: set}
+
+		// Ensure the hit ratio starts at 100% and populate the set such that
+		// number of items specified by the test are expired.
+		assertSetHitRatio(tc, 100)
+		for i := uint32(0); i < test.numExpired; i++ {
+			set.Put(items[i])
+		}
+		mockNow = baseNow.Add(2 * time.Second)
+		for i := test.numExpired; i < test.numItems; i++ {
+			set.Put(items[i])
+		}
+		mockNow = baseNow.Add(test.ttl + time.Second)
+
+		// Ensure checking existence does not modify the hit ratio.
+		for _, itemIdx := range test.itemIdxs {
+			tc.name = fmt.Sprintf("%s-itemidx %d", test.name, itemIdx)
+			set.Exists(items[itemIdx])
+			assertSetHitRatio(tc, 100)
+		}
+
+		// Ensure existence checking does not modify the hit ratio.
```

Comment:
This is already checked above.

Reply:
Meh, not sure how I doubled that up.  Nice catch.  Removed.